music_assistant/server/providers/tidal/__init__.py

```diff
@@ -0,0 +1,393 @@
+    instance_id: str | None = None,  # noqa: ARG001
+    action: str | None = None,
+    values: dict[str, ConfigValueType] | None = None,
+) -> tuple[ConfigEntry, ...]:
+    """
+    Return Config entries to setup this provider.
+
+    instance_id: id of an existing provider instance (None if new instance setup).
+    action: [optional] action key called from config entries UI.
+    values: the (intermediate) raw values for config entries sent with the action.
+    """
+    # config flow auth action/step (authenticate button clicked)
+    if action == CONF_ACTION_AUTH:
+        async with AuthenticationHelper(mass, values["session_id"]) as auth_helper:
+            tidal_session = await tidal_code_login(auth_helper)
+            if not tidal_session.check_login():
+                raise LoginFailed("Authentication to Tidal failed")
+            # set the retrieved token on the values object to pass along
+            values[CONF_AUTH_TOKEN] = tidal_session.access_token
+            values[CONF_REFRESH_TOKEN] = tidal_session.refresh_token
+            values[CONF_EXPIRY_TIME] = tidal_session.expiry_time.isoformat()
+            values[CONF_USER_ID] = str(tidal_session.user.id)
+
+    # return the collected config entries
+    return (
+        ConfigEntry(
+            key=CONF_AUTH_TOKEN,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Authentication token for Tidal",
+            description="You need to link Music Assistant to your Tidal account.",
+            action=CONF_ACTION_AUTH,
+            action_label="Authenticate on Tidal.com",
+            value=values.get(CONF_AUTH_TOKEN) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFRESH_TOKEN,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Refresh token for Tidal",
+            description="You need to link Music Assistant to your Tidal account.",
+            hidden=True,
+            value=values.get(CONF_REFRESH_TOKEN) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_EXPIRY_TIME,
+            type=ConfigEntryType.STRING,
+            label="Expiry time of auth token for Tidal",
+            hidden=True,
+            value=values.get(CONF_EXPIRY_TIME) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_USER_ID,
+            type=ConfigEntryType.STRING,
+            label="Your Tidal User ID",
+            description="This is your unique Tidal user ID.",
+            hidden=True,
+            value=values.get(CONF_USER_ID) if values else None,
+        ),
+    )
+
+
+class TidalProvider(MusicProvider):
+    """Implementation of a Tidal MusicProvider."""
+
+    _tidal_session_manager: TidalSessionManager | None = None
+
+    async def handle_setup(self) -> None:
+        """Handle async initialization of the provider."""
+        async with TidalSessionManager(
+            self.instance_id, self.config, self.mass
+        ) as tidal_session_manager:
```

Comment:
this is a bit weird to use a contextmanager. I was under the impression that you needed to validate the session each call. In that case you can perhaps store the session as attribute on your class attribute and use the contextmanager  on every call that needs the session. Within the context manager you can then perform a quick check if the token needs to be refreshed, cleanup resources etc.

There is no actual need for a contextmanager but its just comfortable. You can even turn a method within your class into a contextmanager: https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager

so you can store/cache the session as instance attribute.
and then in every function that needs access to the session you do something like this:

session = self.get_tidal_session()
and then do something with that session
or the contextmanager approach:

async with self.get_tidal_session() as session:
   session.do_something()

and in the "get_tidal_session" you do all the checks needed if the session is still valid, auth token fresh etc.

You can also create a decorator and decorate all functions.

or the lazy approach is to just call whatever on the todil session and catch the exception that the token expired, refresh it then and then retry.

So enough approaches ;-)

Reply:
To be honest, I think I was going round in circles today.
I already started changing this whole context manager approach as it also seemed completely strange to me.
Should I convert this PR to a draft for the time being?